image-registry/automate-pruning.md

```diff
@@ -0,0 +1,336 @@
+as well as to reduce registry blob storage utilization. Today, pruning is
+configured through a manual, multi-step process.
+
+### Goals
+
+- Provide a one-step process for cluster admins to configure recurring
+  scheduled image pruning.
+- Default pruning to be enabled on an OpenShift cluster at install time for new
+  clusters.
+- Provide alerting if image pruning fails, which may indicate a cluster is at
+  risk of failure.
+
+### Non-Goals
+
+- Fix fundamental issues with the current pruning process (ex: pruning images
+  that may be referenced in a CRD).
+- Enable pruning on existing clusters during upgrade.
+- Provide alerting if a cluster is at risk due to too many images.
+
+## Proposal
+
+### User Stories
+
+#### Prune images in etcd
+
+As an OpenShift cluster admin
+I want the cluster to automatically prune ImageStreams stored in etcd
+So that etcd storage is not overwhelmed by unused imagestreams
+And the cluster remains healthy
+
+#### Prune images on the registry
+
+As an OpenShift cluster admin
+I want the cluster to automatically prune unused images stored in the registry
+So that registry storage is managed
+And I don’t pay for storage consumed by images my organization doesn’t use
+
+#### Orphaned blobs
+
+As an OpenShift cluster admin
+I want the cluster to automatically remove orphaned blobs stored in the registry
+So that registry storage is managed
+And I don’t pay for storage consumed by images my organization doesn’t use
+
+### Implementation Details/Notes/Constraints
+
+The implementation relies on the current image pruning process invoked by the
+`oc adm prune` command. Running this in a container can be accomplished by
+using the `cli` imagestream installed by the samples operator. The image must
+be run using a service account with the `system:image-pruner` cluster role.
+
+The image registry operator will be updated to configure and reconcile all the
+components needed to run image pruning on an OpenShift cluster.
+Specifically, this includes the following:
+
+1. A `ServiceAccount` dedicated to the image pruner.
+2. A `ClusterRoleBinding` granting the `system:image-pruner` role to the pruner
+service account.
+3. A `CronJob` which runs the image pruner with the following tunable
+   parameters:
+   1. Schedule for `CronJob`
+   2. `--keep-tag-revision` command line argument for the pruner
+   3. `--keep-younger-than` command line argument for the pruner
+   4. Resources for the `CronJob` pod template
+
+Installation and configuration of the pruner will be accomplished via a
+separate CRD. These components should be created/reconciled if the pruning CRD
+is created or updated.
+
+When the pruning CRD is deleted, the pruning `CronJob` and its related components
+should also be deleted.
+
+The operator's behavior with respect to managing the pruner is independent of
+the `ManagementState` specified on the image registry operator's
+`ClusterOperator` object.
```

Comment:
if the registry is "unmanaged" you'll still prune both etcd + storage?  I thought you were not going to prune storage in this case.

if the registry is "removed" you can't prune storage, so certainly that changes behavior?

Reply:
We were only going to prune etcd if the registry is `Unmanaged` - I'll clarify this paragraph.

Reply:
@bparees updated